docs/proposals/202105256-cluster-class-and-managed-topologies.md

```diff
@@ -631,7 +698,9 @@ Builtin variables are available under the `builtin.` prefix. Some examples:
               the template and validating if the result is a valid YAML or JSON value.
     - We will only implement syntax validation. A semantic validation is not possible as the patches will be applied 
       to provider templates. **It’s the responsibility of the ClusterClass author to ensure the patches are semantically valid**.
-
+  - `MachineHealthCheck`
+    - If defined at `ClusterClass.spec.machineHealthCheck` must not be defined in `ClusterClass.spec.workers.machineDeployments[i].machineHealthCheck` or `ClusterClass.spec.controlPlane.machineHealthCheck` 
+    - (defaulting) All fields required for MachineHealthChecks, including UnhealthyConditions, are defaulted as described in the API definition.
```

Comment:
I think I would prefer having the same defaulting on the ClusterClass webhook as on the MHC webhook to have the MHC config in the ClusterClass behave the same way as the MHC itself.

Some context:
* We currently default:
  * spec.maxUnhealthy not set => 100%
  * spec.nodeStartupTimeout not set => 10m
  * spec.remediationTemplate != nil && spec.remediationTemplate.namespace empty => spec.remedationTemplate.namespace => MHC.namespace

**If the defaulting behavior is the same**

* nodeStartupTimeout not set => defaulted in the ClusterClass => explicitly set in the MHC
    * => ClusterClass.spec.machineHealthCheck and MHC will be the same

**if we don't have defaulting in the ClusterClass**

* nodeStartupTimeout not set => not defaulted in the ClusterClass, but defaulted in MHC
    *  => MHC will have the timeout, ClusterClass.spec.machineHealthCheck won't
        * we have to implement a logic where we only keep ClusterClass.spec.machineHealthCheck sync with MHC if the fields in ClusterClass.spec.machineHealthCheck is not nil.

**What happens if the MHC defaulting changes**

We have the weird behavior that the MHC spec depends on when it has been created by the cluster topology controller, because the fields are "unmanaged" from a topology controller perspective. I.e. old MHC's will have the fields filled by the old default logic, new MHCs by the new. Depending on how the defaulting changes we might end up with a mixture.

When the ClusterClass itself defaults the defaults are persisted at the time of ClusterClass creation in etcd and are then consistently used.

tl;dr I lean towards keeping the defaulting logic sync (like we did in cases like KCP vs KCPTemplate) to avoid weird effects.

Reply:
> If the MHC controller takes care of handling the default behavior ... become a hidden behavior for the user.

I don't think this is a hidden behavior, but a feature of ClusterClass explicitly documented in [template<->object-reconciliation](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/202105256-cluster-class-and-managed-topologies.md#notes-on-template---object-reconciliation).

ClusterClass as of today takes care of enforcing a _subset of fields_ in the generated objects, but the controller/the users are responsible for the missing ones, thus opening up to things like user customization or the controllers to store some information in the spec (e.g. for making them survive move operations).

And if you think about it is the same behaviour users get when they use kubectl apply (all the things the were explicit about, are preserved, everything else is delegated to the system that eventually apply defaults when things are created)

Introducing explicit defaulting for MHC will mean that ClusterClass, kind of goes in the opposite direction, so I'm -1 for this

Reply:
> Introducing explicit defaulting for MHC will mean that ClusterClass, kind of goes in the opposite direction

From that point of view it makes sense. So the main idea of ClusterClass in that respect is to preserve user intent (and don't overwrite MHC webhook intent). 

That means we should be very very conservative with any defaulting we do on ClusterClass. But this fits the current behavior as we're only defaulting namespaces right now.

Reply:
I've created a parallel issue for defaulting at https://github.com/kubernetes-sigs/cluster-api/issues/5769 and removed the specifics of where defaulting will happen from the proposal to make it an implementation detail.

Reply:
Sounds good, can be resolved for now from my side.